### DISCUSSION TITLE
Connect Variable to use for GPIO signal?

### DISCUSSION THREAD

**Discussion by la-tulipe** [dtcooper/raspotify#455] [Q&A] [answered]
Hi There!
Thanks for this great project. I have been using it for a couple of years now without any issues.
I would like to be able to use a variable to put on one of the GPIO pins to light-up a LED or a relay (TBD) upon (dis)connection.
I guess it exists as in the spotify app we see the connect icon turn green.
Any idea where I can find that and how to use it?
Thanks,
Oscar

**Comment by JasonLG1979** [comment, marked as answer]
What you're looking for is the `--onevent` librespot flag.

https://github.com/librespot-org/librespot/wiki/Options

https://github.com/librespot-org/librespot/issues/185

https://github.com/librespot-org/librespot/blob/dev/src/player_event_handler.rs

**Reply by la-tulipe** [reply]
That was fast!
Thank you very much :-)

**Reply by JasonLG1979** [reply]
The systemd unit file has been security hardened for the next release so you may need to modify it with a drop-in to make the script visible and executable by the Raspotify service.  

`sudo systemctl edit raspotify`

https://unix.stackexchange.com/questions/398540/how-to-override-systemd-unit-file-settings

That is uncharted territory so needless to say it is an undocumented use case so when/if you do get it working please report back so it can be documented for use by other users.

**Reply by la-tulipe** [reply]
Thanks! Sorry for the delay... Hope to get to it durng the weeks ahead...

**Reply by JasonLG1979** [reply]
> Thanks! Sorry for the delay... Hope to get to it durng the weeks ahead...

Keep in mind that the `onevent` script will be executed with the same privileges as the service which are basically nothing and most of the file system is either read only or completely inaccessible to the Raspotify service. Putting `fifo`'s  under `/var/run` seems to work so I'd imagine that would be a good place to start. Otherwise if need be I can open a specific location like `/var/run/raspotify` for `onevent` scripts? I'd need to figure out where the most proper place would be according to convention and the LSB.

Again this is uncharted territory.

**Reply by JasonLG1979** [reply]
Speaking of `fifo`'s, A not so elegant solution could be:

Program B creates and listens to `/var/run/programB/fifo` and does whatever you want to do.
`onevent` script writes to `/var/run/programB/fifo`.

**Reply by JasonLG1979** [reply]
Totally untested but based on [player_event_handler.rs](https://github.com/librespot-org/librespot/blob/dev/src/player_event_handler.rs) (the part of librespot that talks to the `onevent` scripts) a skeleton of a `python` `onevent` script might look something like this:

```python3
import os

if os.environ['PLAYER_EVENT'] == 'changed':
    old_track_id = os.environ['OLD_TRACK_ID']
    new_track_id = os.environ['TRACK_ID']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'started':
    track_id = os.environ['TRACK_ID']
    # do suff

elif os.environ['PLAYER_EVENT'] == 'stopped':
    track_id = os.environ['TRACK_ID']
    # do stuff
   
elif os.environ['PLAYER_EVENT'] == 'playing':
    track_id = os.environ['TRACK_ID']
    track_duration_ms = os.environ['DURATION_MS']
    track_position_ms = os.environ['POSITION_MS']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'paused':
    track_id = os.environ['TRACK_ID']
    track_duration_ms = os.environ['DURATION_MS']
    track_position_ms = os.environ['POSITION_MS']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'preloading':
    track_id = os.environ['TRACK_ID']
    # do stuff

elif os.environ['PLAYER_EVENT'] == 'volume_set':
    volume = os.environ['VOLUME']
    # do stuff
```

Also note that the script will be called in a non-blocking fashion.

**Reply by JasonLG1979** [reply]
Writing to a fifo might look something like this in Python3:

```python3
#!/usr/bin/python3
import os
import json

fifo_path = "./my_fifo"

try:
    os.mkfifo(fifo_path, 0o666)
except FileExistsError:
    # That's Ok...
    pass

except PermissionError as e:
    # You don't have permission to create a file in that location...
    print("PermissionError: {}".format(e))
    pass

except Exception as e:
    # Handle this...
    print(e)
    pass

json_dict = None

if os.getenv('PLAYER_EVENT') == 'changed':
    json_dict = {
        'event': os.getenv('PLAYER_EVENT'),
        'old_track_id': os.getenv('OLD_TRACK_ID') or '',
        'new_track_id': os.getenv('TRACK_ID') or ''
    }

elif os.getenv('PLAYER_EVENT') in ('started', 'stopped', 'preloading'):
    json_dict = {
        'event': os.getenv('PLAYER_EVENT'),
        'track_id': os.getenv('TRACK_ID') or ''
    }

elif os.getenv('PLAYER_EVENT') in ('playing', 'paused'):
    json_dict = {
        'event': os.getenv('PLAYER_EVENT'),
        'track_id': os.getenv('TRACK_ID') or '',
        'duration_ms': os.getenv('DURATION_MS') or '0',
        'position_ms': os.getenv('POSITION_MS') or '0'
    }

elif os.getenv('PLAYER_EVENT') == 'volume_set':
    json_dict = {
        'event': os.getenv('PLAYER_EVENT'),
        'volume': os.getenv('VOLUME') or '0'
    }

if json_dict:
   print(json_dict)
   with open(fifo_path, 'w') as f:
        try:
            json.dump(json_dict, f)
        except KeyboardInterrupt:
            pass
        except Exception as e:
            # Handle this...
            print(e)
            pass

```
